File: blind75/greedy/53-maximumSubarray.js
// 53. Maximum Subarray
// Given an integer array nums, find the contiguous subarray (containing at least one number) which has the largest sum and return its sum.
// A subarray is a contiguous part of an array.


/**
 * @param {number[]} nums
 * @return {number}
 */
 var maxSubArray = function(nums) {
    let maxSum = nums[0];
    let currSum = nums[0];
    for (let i = 1; i < nums.length; i++) {
        // if currSum < 0, we will left the sum above because it will not help
        // to make the result bigger 
        if (currSum < 0) {
            currSum = 0;
        }
        // currSum means the maxSum till the index 
        currSum = currSum + nums[i];
        maxSum = Math.max(maxSum, currSum);
    }
    return maxSum;
};


// 这是我自己想出来的方法 很多可以优化的空间
// var maxSubArray = function(nums) {
//     let sum = new Array(nums.length).fill(nums[0]);
//     for (let i = 1; i < nums.length; i++) {
//         sum[i] = Math.max(sum[i - 1] + nums[i], nums[i])
//     }
//     return Math.max(...sum);
// };

let nums = [-2,1,-3,4,-1,2,1,-5,4];
let nums1 = [1];
let nums2 = [5,4,-1,7,8]
console.log(maxSubArray(nums));
console.log(maxSubArray(nums1));
console.log(maxSubArray(nums2));

File: specified-learning专项训练/greedy/advanced进阶/53-maximumSubarray.py
# 53. Maximum Subarray
# Given an integer array nums, find the contiguous subarray (containing at least one number) which has the largest sum and return its sum.
# A subarray is a contiguous part of an array.

# 从第二位元素遍历 如果前面最大和为正数，则相加，不然就不加
# 最大值则为自己
from typing import List
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        maxSum = [nums[0] for i in range(len(nums))]
        for i in range(1, len(nums)):
            if maxSum[i - 1] > 0:
                maxSum[i] = maxSum[i - 1] + nums[i]
            else:
                maxSum[i] = nums[i]
        return max(maxSum)

nums = [-2,1,-3,4,-1,2,1,-5,4]
print(Solution().maxSubArray(nums))
nums = [1]
print(Solution().maxSubArray(nums))
nums = [5,4,-1,7,8]
print(Solution().maxSubArray(nums))